### ISSUE TITLE
Firehed\U2F\InvalidDataException Missing key cid_pubkey

### ISSUE THREAD

**Issue by goncalo-ribeiro** [Firehed/u2f-php#14]
Hi, I'm trying to implement this library into my laravel project and whenever i try to register a response:

`$resp = RegisterResponse::fromJson($u2fSignatureJson ?? '');`

I get a `Firehed\U2F\InvalidDataException Missing key cid_pubkey` exception.
I used the examples given as a guiding line but I haven't been able to fix it, below is the function where this command is called 

```
 public function signature(Request $request){
        $user = Auth::user();
        $u2fRequestJson = $request->input('jsonRequest');
        $u2fSignatureJson = $request->input('jsonSignature');      
        
        $this->server->setRegisterRequest(Session::get('request'));

        $u2fSignature = json_decode($u2fSignatureJson);
        $u2fSignature->cid_pubkey = 'unused';
        $u2fSignatureJson = json_encode($u2fSignature);
        
        $resp = RegisterResponse::fromJson($u2fSignatureJson ?? '');
            

        // Attempt to register with parsed response
        $registration = $this->server->register($resp);
...
```
The `request` object received by this function is the laravel equivalent of `$_POST["form_field"]` and has nothing to do with this library's `$server->generateRegisterRequest();`.
I really would appreciate any help you could spare me, since I wasn´t able to find any info about this problem elsewhere.

**Comment by Firehed** [comment]
Hi,

I assume this bit was attempting to work around the issue?
```php
$u2fSignature = json_decode($u2fSignatureJson);
$u2fSignature->cid_pubkey = 'unused';
$u2fSignatureJson = json_encode($u2fSignature);
 ```

You'll absolutely need to remove that, since it will always cause a signature mismatch. However that's not (likely) the source of the issue.

Can you send the raw POST data? Either `print_r($_POST)` or something out of the network inspector from your browser should suffice. It shouldn't contain any sensitive information, only a public key. Additionally, can you indicate what browser you're using and what vendor's U2F key you're trying to test with?

Thanks!

---

On a related note, you may want to hold off until #11 lands, which should be fairly soon. That should result in much better cross-browser support going forward.

**Comment by goncalo-ribeiro** [comment]
I apologize for the delay, but I currently don't have access to a U2F key, I'm using my friend's key at the moment.
That code block you quoted is indeed a try at a workaround, sry it slipped.
I did a ``dd($_POST)`` and I got the following array:
```
array:3 [
  	"_token" => "sir6SNhTsMR0Hwn7r8PZShxG7iHTAcmQ6d2nmn9A"
  	"jsonRequest" => "{"version":"U2F_V2",
  		"challenge":"4YogTFhK_AxCTPCUpAKLPg",
  		"appId":"https://squirrel.test"}"
  	"jsonSignature" => "{"registrationData":"BQTpfFv-_4e8XeFNdQT5P4UYRJXmECWVoPSM4BKQYKib_3N0T_SY7wHZavNb7GGnGq1Ui4ospnt0f_BcElTRXTiFQIJiUkLlU963WkuaevOyDmzBHtLPuHY85AE_fZmriIJCHKfhTewl0I5nXRpLAI1A7gfWyZCtmx4_sp8kBwSJ9VMwggE1MIHcoAMCAQICCwCuQTk0cvzUBry8MAoGCCqGSM49BAMCMBUxEzARBgNVBAMTClUyRiBJc3N1ZXIwGhcLMDAwMTAxMDAwMFoXCzAwMDEwMTAwMDBaMBUxEzARBgNVBAMTClUyRiBEZXZpY2UwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAATjAVf_Zlzb-XTljolBwW5oe24071h3ymiXj-puuNyLEW0ISxC-nL6SfOx9bDUNRybNZOJsmL2AM1TwzaU898o1oxcwFTATBgsrBgEEAYLlHAIBAQQEAwIFIDAKBggqhkjOPQQDAgNIADBFAiEAwaOmji8WpyFGJwV_YrtyjJ4D56G6YtBGUk5FbSwvP3MCIAtfeOURqhgSn28jbZITIn2StOZ-31PoFt-wXZ3IuQ_eMEUCIQCULLuAQfzQ5ke7AtH3UwcM_7llD8h3I6pS-XXZG5uS-gIgE2oLtfF9WcjNw6u0-g7dOilGOY6WaVhuwhGVmhbUoCw",
  		"version":"U2F_V2",
  		"challenge":"4YogTFhK_AxCTPCUpAKLPg",
  		"appId":"https://squirrel.test",
  		"clientData":"eyJjaGFsbGVuZ2UiOiI0WW9nVEZoS19BeENUUENVcEFLTFBnIiwib3JpZ2luIjoiaHR0cHM6Ly9zcXVpcnJlbC50ZXN0IiwidHlwIjoibmF2aWdhdG9yLmlkLmZpbmlzaEVucm9sbG1lbnQifQ"}
]
```
I'm using a yubikey and the latest version of chrome also.
Thanks in advance

**Comment by Firehed** [comment]
Hm, that value certainly isn't in there. Looking over the spec it does appear that field can be treated as optional, and maybe the latest hardware rev on the key no longer emits that field.

Are you using the exact version of `u2f-api.js` linked in the readme? (https://github.com/google/u2f-ref-code/blob/master/u2f-gae-demo/war/js/u2f-api.js)

There's a good chance this gets fixed in 2.0 (via #11) just due to a bunch of changes to data parsing, but I can't easily reproduce this issue to verify. I'll reiterate that it's probably better to hold off until that lands before going too deep with integrating this, since it will allow cross-browser support.

**Comment by bowsernet** [comment]
Hi, I'm having the same issue. Actually, it's very easy to reproduce, just use your own example in latest Chrome (yes, I think the browser might actually be the underlying issue here since I haven't changed a thing in my project and suddenly the registrations stopped working)
https://github.com/Firehed/u2f-php-examples

I'd love to help here but without deeper knowledge of how this works low level and with no time on my hands, I'm no use unfortunately.

Thanks for looking into this!!!

**Comment by Firehed** [comment]
I appreciate the info - I'll try to dig in and figure it out. I can now also reproduce the issue in the latest Chrome, maybe my other machine hadn't updated yet.

**Comment by Firehed** [comment]
@bowsernet @goncalo-ribeiro Can you trying running the exact version `d5fbb9a` (#20) and see if that resolves the issue? I don't have a great way to integration test the legacy u2f-api.js stuff at the moment since it requires true https even over localhost (unlike WebAuthn, which has an exception for it). 

If it passes a sanity-check I will try to get it integrated properly as a new tag with proper test coverage.

**Comment by bowsernet** [comment]
@Firehed fix worked. Thanks a lot for such a quick response!

**Comment by Firehed** [comment]
@bowsernet @goncalo-ribeiro This is fixed on the v1 branch, and I just published `1.0.1` from said branch.

I'm leaving this open for now since merging the fixes back into master as part of the WebAuthn updates coming in 2.0 is a bit tricky, and I don't want to introduce a regression.

Make sure you read the comments on [that change](https://github.com/Firehed/u2f-php/commit/38780b93b8bb9d6cf7c20536782b030cbf3c3abb) since there is still some weird behavior in Chrome that I experienced in my testing (the certificate for the key itself is dated and only valid on 2000-01-01). I don't know the best way to get that information back to the Chrome team, but if you're using the strict verification of key providers (with `Server::setTrustedCAs()`) you're likely to still experience issues. That one's outside of my control unfortunately.
